### PR TITLE
Adding French

### PR DESCRIPTION
**Pull request by rfrerebe**
Failing on
> Ploeh.Numsense.Numeral.tryParseFrench "quatre-cent-quatre-vingt-dix";;
val it : int option = Some 490
> Ploeh.Numsense.Numeral.tryParseFrench "quatrecentquatrevingtdix";;
val it : int option = Some 434

still some work to do ...

### CODE REVIEW

**Comment by ploeh** [comment]
@rfrerebe, thank you for your interest in contributing to Numsense.

> still some work to do

Since this is the case, I'll hold off with initiating a review until it's ready. When it's ready, please add a note here to that effect, since GitHub doesn't send me any notifications when a pull request branch is updated.

**Comment by rfrerebe** [comment]
Hello Mark,

Thanks for tour answer.
For your info, property based testing is fine while manual unit test is
not.
I was surprised by that.

May I ask you to add a Licence.md to your project?
I was surprised by cla email.
What is the purpose of this project ?
I might be wrong but I found the cla very corporate and not very OSS
friendly.

Regards

Rémy



Le mer. 6 janv. 2016 à 08:13, Mark Seemann <notifications@github.com> a
écrit :

**Comment by ploeh** [comment]
> May I ask you to add a Licence.md to your project?

I'm not sure the *CLAassistant* GitHub webhook allows that; it's driven by a gist.

> What is the purpose of this project?

There's no particular purpose. I did the original languages (English and Danish) because I was surprised to find that there wasn't such a library available (that I could easily find). From there, it just turned out that it was fun to do :smile: 

> I might be wrong but I found the cla very corporate and not very OSS friendly.

I took the CLA originally suggested by *CLAassistant* and edited it to remove the worst. Feel free to suggest a more OOS-friendly CLA...

That said, the code and software itself is released under the MIT licence, as you may already have noticed.

**Comment by ploeh** [comment]
I've now added a section about the CLA to the CONTRIBUTING.md document, in order to give advance warning to potential contributors.

**Comment by rfrerebe** [comment]
Hello,

Thanks for your answer.
MIT is fine.

I did not saw it before...
I was looking for LICENCE.md upper case I guess and I missed your file.
Anyway I was tired yesterday evening....

I found a solution to the bug I had.
I won't have much time before this week end to update the last unit test
part.

I will keep you posted.
CLA signed

Rgds

Rémy


Le mer. 6 janv. 2016 à 21:27, Mark Seemann <notifications@github.com> a
écrit :

**Comment by ploeh** [comment]
> I was looking for LICENCE.md upper case I guess and I missed your file.

That's good feedback, thank you. I've now renamed it to upper-case letters, in order to make it more visible.

**Comment by rfrerebe** [comment]
Seems ok for me.
Please have look.

Thanks

**Comment by ploeh** [comment]
Thank you for your interest in contributing to Numsense. FTR, I'm currently travelling, and have only limited time to evaluate pull requests. In due time, I'll work through my backlog, but it may take weeks :flushed: :sweat:

**Comment by rfrerebe** [comment]
Sure, :-)
I saw on Twitter that on you were doing a conference in London.
I hope it went well.

It was fun to work on Numsense.
I found it was a good exercise.
French was tough because some corner cases exist for plural
("quatre-vingt(s)" and "cent(s)" mostly).

I will switch to other projects.

Good luck for you backlog.
It's good to see Numsense attracting many pull request.

Rémy

File: Numsense.UnitTests/FrenchExample.fs [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+let ``toFrench returns correct result`` (i, expected) =
+    let actual =
+        i
+        |> Numeral.toFrench
+    expected =! actual
+    
+    let i' = 
+        actual
+        |> Numeral.tryParseFrench
+    Some i =! i'
```
Comment (ploeh) [inline Numsense.UnitTests/FrenchExample.fs]: Why is this call to `tryParseFrench` and this assertion here?

Comment (rfrerebe) [inline Numsense.UnitTests/FrenchExample.fs]: It's to make sure there is no issue to "round-trip" properly.
(ie : a = a |>  Numeral.toFrench |> Numeral.tryParseFrench ) 

It tests string with dash. The other test is taking mostly string whitout dash in input.

Do you prefer I remove it ?

Comment (ploeh) [inline Numsense.UnitTests/FrenchExample.fs]: Don't you already have that property covered with `Ploeh.Numsense.NumeralProperties.tryOfFrench is the inverse of toFrench`?

**Comment by rfrerebe** [comment]
Yes, I updated the pull request

File: Numsense.UnitTests/FrenchExample.fs (hunk 1 above)
Comment (rfrerebe) [inline Numsense.UnitTests/FrenchExample.fs]: Hello,

I pushed a new version. Is it ok for you ?

Rémy
Le dim. 24 janv. 2016 à 17:30, Mark Seemann <notifications@github.com> a
écrit :

**Comment by Pvlerick** [comment]
@ploeh seeing this pull request, how would you handle localized versions of languages (I'm thinking of Belgian French but there are other examples).

**Comment by ploeh** [comment]
> how would you handle localized versions of languages

We had a [discussion about that related to the Portuguese pull request](https://github.com/ploeh/Numsense/pull/18/files#r50638731). Perhaps we could use the same approach here?

I'll need to native speakers to guide me, because I suspect that every situation is going to be slightly different than the last. After all, we're dealing with the real world here, so no consistent rules are guaranteed to always hold.

**Comment by Pvlerick** [comment]
@ploeh didn't see that, thanks. I guess ```toBelgianFrench``` would do it then.

**Comment by ploeh** [comment]
It builds and runs all tests without warnings on my machine, so from a technical perspective I think we're good to go :+1:

As I don't read French, I'd like someone who does to review that part of it, if at all possible. I don't expect any errors, but it's always good with a second pair of eyes :smile:

**Comment by ploeh** [comment]
I solicited a review on Twitter: https://twitter.com/ploeh/status/693905655799943168

Please retweet and spread the word :smile:

**Comment by Pvlerick** [comment]
Zero (0) is spelled with an acute accent: "zéro".

**Pull request closed without merging**